modules/openapi-generator/src/main/resources/python/python-experimental/signing.mustache

```diff
@@ -0,0 +1,317 @@
+        :param headers: A dict containing the HTTP request headers.
+        :param body: The string representation of the HTTP request body.
+        :param query_params: A string representing the HTTP request query parameters.
+        :return: A tuple containing two dict objects:
+            The first dict contains the HTTP headers that are used to calculate the HTTP signature.
+            The second dict contains the HTTP headers that must be added to the outbound HTTP request.
+        """
+
+        if body is None:
+            body = ''
+        else:
+            body = json.dumps(body)
+
+        # Build the '(request-target)' HTTP signature parameter.
+        target_host = urlparse(host).netloc
+        target_path = urlparse(host).path
+        request_target = method.lower() + " " + target_path + resource_path
+        if query_params:
+            raw_query = urlencode(query_params).replace('+', '%20')
+            request_target += "?" + raw_query
+
+        # Get current time and generate RFC 1123 (HTTP/1.1) date/time string.
+        now = datetime.datetime.now()
+        stamp = mktime(now.timetuple())
+        cdate = formatdate(timeval=stamp, localtime=False, usegmt=True)
+        # The '(created)' value MUST be a Unix timestamp integer value.
+        # Subsecond precision is not supported.
+        created = self.get_unix_time(int(now))
+        if self.signature_max_validity is not None:
+            expires = self.get_unix_time(now + self.signature_max_validity)
+
+        signed_headers_dict = {}
+        request_headers_dict = {}
+        for hdr_key in self.signed_headers:
+            hdr_key = hdr_key.lower()
+            if hdr_key == HEADER_REQUEST_TARGET:
+                value = request_target
+            elif hdr_key == HEADER_CREATED:
+                value = '{0}'.format(created)
+            elif hdr_key == HEADER_EXPIRES:
+                value = '{0}'.format(expires)
+            elif hdr_key == HEADER_DATE:
+                value = cdate
+                request_headers_dict[HEADER_DATE] = '{0}'.format(cdate)
+            elif hdr_key == HEADER_DIGEST:
+                request_body = body.encode()
+                body_digest, digest_prefix = self.get_message_digest(request_body)
+                b64_body_digest = b64encode(body_digest.digest())
+                value = digest_prefix + b64_body_digest.decode('ascii')
+                request_headers_dict[HEADER_DIGEST] = '{0}{1}'.format(
+                    digest_prefix, b64_body_digest.decode('ascii'))
+            elif hdr_key == HEADER_HOST:
+                value = target_host
+                request_headers_dict[HEADER_HOST] = '{0}'.format(target_host)
+            else:
+                value = headers[hdr_key]
+            signed_headers_dict[hdr_key] = value
+
+        return signed_header_dict, request_headers_dict
+
+    def get_message_digest(self, data):
```

Comment:
Can change this method's name to add a `__` prefix so we can tell that it is not our publicly used method?
See http://python.robasworld.com/private-member-name-mangling/

Reply:
done